resilience4j-retrofit/src/main/java/io/github/resilience4j/retrofit/RetrofitRateLimiter.java

```diff
@@ -0,0 +1,70 @@
+/*
+ *
+ *  Copyright 2017 Christopher Pilsworth
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *
+ */
+package io.github.resilience4j.retrofit;
+
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.ratelimiter.RateLimiter;
+import io.github.resilience4j.retrofit.internal.DecoratedCall;
+import javaslang.control.Try;
+import okhttp3.MediaType;
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.Response;
+
+import java.io.IOException;
+
+/**
+ * Decorates a Retrofit {@link Call} to check with a {@link RateLimiter} if a call can be made.
+ * Returns an error response with a HTTP 429 (too many requests) code and a message which indicates that the client
+ * prevented the request.
+ *
+ * <p>
+ * <code>
+ * RetrofitRateLimiter.decorateCall(rateLimiter, call);
+ * </code>
+ */
+public interface RetrofitRateLimiter {
+
+    int ERROR_CODE = 429;
+    String ERROR_MESSAGE = "Too many requests for the client";
+
+    /**
+     * Decorate {@link Call}s allow {@link CircuitBreaker} functionality.
+     *
+     * @param rateLimiter  {@link RateLimiter} to apply
+     * @param call            Call to decorate
+     * @param <T> Response type of call
+     * @return Original Call decorated with CircuitBreaker
+     */
+    static <T> Call<T> decorateCall(final RateLimiter rateLimiter, final Call<T> call) {
+        return new DecoratedCall<T>(call) {
+            @Override
+            public Response<T> execute() throws IOException {
+                Try.CheckedSupplier<Response<T>> restrictedSupplier = RateLimiter.decorateCheckedSupplier(rateLimiter, call::execute);
```

Comment:
I have updated the code to this now:
```
public Response<T> execute() throws IOException {
                final Try.CheckedSupplier<Response<T>> restrictedSupplier = RateLimiter.decorateCheckedSupplier(rateLimiter, call::execute);
                return Try.of(restrictedSupplier)
                        .recover(ex -> Match(ex).of(
                                Case(instanceOf(RequestNotPermitted.class), tooManyRequestsError()),
                                Case(instanceOf(IllegalStateException.class), tooManyRequestsError()))
                        ).get();
            }
```
However, Try seems to want to always swallow the exceptions and return some Response<T>.  I really only want to consume the `RequestNotPermitted` and perhaps `IllegalStateException` exceptions so as not to change the interface for `Call` in decorating.  

The code above will give a MatchError on the a IOException from Call.execute - I would like to re-throw on the "else" match if possible.  Is there any way of doing this?